File: 双指针/剑指05替换空格.py
class Solution:
    def replaceSpace(self, s: str) -> str:  
        counter = s.count(' ')        
        res = list(s)
        # 每碰到一个空格就多拓展两个格子，1 + 2 = 3个位置存’%20‘
        res.extend([' '] * counter * 2)       
        # 原始字符串的末尾，拓展后的末尾
        left, right = len(s) - 1, len(res) - 1        
        while left >= 0:
            if res[left] != ' ':
                res[right] = res[left]
                right -= 1
            else:
                # [right - 2, right), 左闭右开
                res[right - 2: right + 1] = '%20'
                right -= 3
            left -= 1
        return ''.join(res)

File: 链表/剑指05替换空格.java
class Solution {
    public String replaceSpace(String s) {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < s.length(); i++){
            if(s.charAt(i) != ' '){
                result.append(s.charAt(i));
            }else{
                result.append("%20");
            }
        }
        return result.toString();
    }
}

class Solution:
    def replaceSpace(self, s: str) -> str:
        res = []
        for c in s:
            if c == ' ': res.append("%20")
            else: res.append(c)
        return "".join(res)
